scrapy/core/downloader/handlers/http11.py

```diff
@@ -382,6 +382,27 @@ def _cb_latency(self, result, request, start_time):
         return result
 
     def _cb_bodyready(self, txresponse, request):
+        headers_received_result = self._crawler.signals.send_catch_log(
+            signal=signals.headers_received,
+            headers=Headers(txresponse.headers.getAllRawHeaders()),
+            request=request,
+            spider=self._crawler.spider,
+        )
+        for handler, result in headers_received_result:
+            if isinstance(result, Failure) and isinstance(result.value, StopDownload):
+                logger.debug("Download stopped for %(request)s from signal handler %(handler)s",
+                             {"request": request, "handler": handler.__qualname__})
+                with suppress(AttributeError):
```

Comment:
Do you know in which circumstances this happens? Maybe it would be worth indicating in a comment here.

Reply:
To be honest, I didn't dig too much into this part. I would much rather do `Transport.loseConnection`, which [does this check](https://github.com/twisted/twisted/blob/twisted-18.4.0/src/twisted/web/_newclient.py#L1359-L1365) and saves me one call to a private member, but it's not available in the minimum Twisted version we currently support (17.9.0)

Reply:
https://github.com/scrapy/scrapy/issues/4991